dataaccess.js

```diff
@@ -364,18 +379,111 @@ function getPlayerRankings(playerId, showPersonalRank) {
     });
 }
 
-function calculateAllStats() {
-    debug('Calculating stats for every player');
+//This could be a bit dicey, so beware
+function updateResults(gameData) {
+    debug('Updating results for finished game');
+
+    for (let player of gameData.playerRank) {
+        if (player === 'ai') {
+            continue;
+        }
+
+        if (!stats[player]) {
+            //first time player
+            stats[player] = {
+                games: 0,
+                rank: 0,
+                playerName: '',
+                wins: 0,
+                winsAI: 0,
+                percent: 0
+            }
+        }
+
+        if (gameData.playerRank[0] === player) {
+            //This was the winner
+            stats[player].wins++;
+            if (gameData.humanPlayers < 2) {
+                //This was an ai game so it counts towards ai wins
+                stats[player].winsAI++;
+            }
+        }
+
+        stats[player].games++;
+        stats[player].percent = Math.floor(100 * stats[player].wins / stats[player].games);
+
+        //Make sure we have the player's name if it was a new player
+        if (!stats[player].playerName) {
+            return treasonDb.get(player)
+                .then(function (document) {
+                    stats[player].playerName = document.name;
+                }).catch(function (error) {
+                    console.error('Failed to retrieve player: ' + player);
+                    console.error(error);
+                });
+        }
+    }
+
+    //now reorder ranks
+    recalculateRanks(stats);
+
+    //Persist the whole stats - seems silly to do this rather than recalculating on startup
+    //This can be async
+    treasonDb.save(gameStatsDocumentId, {
```

Comment:
It is likely that there will be a race with several different versions of the game stats doc being saved at one time. It would be better to push them into an array then set an immediate timeout to save everything in the queue. That way there would never be a save conflict and the correct order of saves would be enforced.